src/ctap/crypto_wrapper.rs

```diff
@@ -117,24 +160,43 @@ impl PrivateKey {
         ecdsa::SecKey::from_bytes(array_ref!(bytes, 0, 32)).map(PrivateKey::from)
     }
 
+    #[cfg(feature = "with_ed25519")]
+    pub fn new_ed25519_from_bytes(bytes: &[u8]) -> Option<Self> {
+        if bytes.len() != 32 {
+            return None;
+        }
+        if let Ok(secret) = ed25519_dalek::SecretKey::from_bytes(bytes) {
+            let public = ed25519_dalek::PublicKey::from (&secret);
+            Some(Self::Ed25519(ed25519_dalek::Keypair{secret, public}))
+        } else {
+            None
+        }
+    }
+
     /// Returns the corresponding public key.
     pub fn get_pub_key(&self) -> CoseKey {
         match self {
             PrivateKey::Ecdsa(ecdsa_key) => CoseKey::from(ecdsa_key.genpk()),
+            #[cfg(feature = "with_ed25519")]
+            PrivateKey::Ed25519(ed25519_keypair) => CoseKey::from(ed25519_keypair.public),
         }
     }
 
     /// Returns the encoded signature for a given message.
     pub fn sign_and_encode(&self, message: &[u8]) -> Vec<u8> {
         match self {
             PrivateKey::Ecdsa(ecdsa_key) => ecdsa_key.sign_rfc6979::<Sha256>(message).to_asn1_der(),
+            #[cfg(feature = "with_ed25519")]
+            PrivateKey::Ed25519(ed25519_keypair) => ed25519_keypair.try_sign(message).unwrap().to_bytes().to_vec(),
```

Comment:
Under what conditions can signing fail? I'd prefer either making this function return an error, or retrying the signing operation if it's probabilistic. `unwrap` would panic.

Reply:
try_sign() returns Result to be able to signal errors when singing with external signers, such as external HSM. Software implementation in ed25519_dalek::Keypair always returns Ok(_)